### PR TITLE
Fix: topologyAware does not pick up failure domains.

### LINKED ISSUE #3430: topologyAware functionality not working
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

Setting topologyAware to true inside specs section of a cluster.yaml does not result in a cluster that has OSDs under zone or region buckets in the CRUSH map. As such, pools can be created that map replicas or EC chunks to a zone or region osd-failure-domain.

One problem is the 'zone' type is missing from the crushmap rook-ceph is feeding the cluster (see below). The secondary issue is that the osd_location is not being populated with zone and/or region buckets for each OSD.

**Expected behavior:**

Setting topologyAware to true inside specs section of a cluster.yaml should result in a cluster that has OSDs under zone or region buckets in the CRUSH map. Then pools cannot be created that map replicas or EC chunks to a zone or region osd-failure-domain.

**How to reproduce it (minimal and precise):**

1. Provision a three worker OpenShift cluster in EC2 such that each worker is in a distinct availability zone. 
2. Ensure cloud-provided is properly configured such that failure-domain.beta.kubernetes.io/{zone,region} labels are being properly applied to workers.
3. Deploy rook-ceph and create a cluster using a cluster.yaml that contains toplogyAware set to true in the specs.
4. Wait for cluster to finish provisioning
5. Create toolbox pod, rsh to it, and inspect the crushmap to see missing bucket type, and absence of zone or region buckets for pods running in workers with zone/region labels.

**Environment**:
* OS (e.g. from /etc/os-release): Red Hat Enterprise Linux CoreOS 410.8.20190520.0 (Ootpa)
* Kernel (e.g. `uname -a`): 4.18.0-80.1.2.el8_0.x86_64
* Cloud provider or hardware configuration: 3x m4.large masters, 3x m5d.4xlarge workers (in distinct AZs), 
* Rook version (use `rook version` inside of a Rook Pod):

 rook: v1.0.0-187.gf0b0f0e  (f0b0f0e4f31140cc4b81453f1d918d988021db21)

* Kubernetes version (use `kubectl version`):

Client Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.4+cb455d664", GitCommit:"cb455d664", GitTreeState:"clean", BuildDate:"2019-05-19T23:03:23Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"linux/amd64"}

* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):

OpenShift

* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)):

HEALTH_WARN crush map has legacy tunables (require firefly, min is hammer)

**Other Information**

Rook-ceph operator log showing rook-ceph setting initial crushmap. I think this is overriding the upstream version that has the 'zone' crush type. Compare:

https://github.com/rook/rook/blob/1bf3cfa4bf6c19d350f6a3e02bc4846708e17898/pkg/daemon/ceph/client/crush.go#L50

With:

https://github.com/ceph/ceph/pull/27070/files

```
2019-07-08 21:20:54.221869 I | op-cluster: creating initial crushmap
2019-07-08 21:20:54.222002 I | exec: Running command: crushtool -c /tmp/264578157 -o /tmp/614087400
2019-07-08 21:20:54.252555 I | exec: Running command: ceph osd setcrushmap -i /tmp/614087400 --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json --out-file /tmp/281646887
2019-07-08 21:20:55.132306 I | exec: 2
2019-07-08 21:20:55.132424 I | op-cluster: created initial crushmap
```

Crush map on freshly created rook-ceph cluster (note the missing zone type):
```
sh-4.2# cat crush.txt
# begin crush map
tunable choose_local_tries 0
tunable choose_local_fallback_tries 0
tunable choose_total_tries 50
tunable chooseleaf_descend_once 1
tunable chooseleaf_vary_r 1
tunable straw_calc_version 1

# devices
device 0 osd.0 class ssd
device 1 osd.1 class ssd
device 2 osd.2 class ssd
device 3 osd.3 class ssd
device 4 osd.4 class ssd
device 5 osd.5 class ssd

# types
type 0 osd
type 1 host
type 2 chassis
type 3 rack
type 4 row
type 5 pdu
type 6 pod
type 7 room
type 8 datacenter
type 9 region
type 10 root

# buckets
host ip-10-0-139-184 {
	id -3		# do not change unnecessarily
	id -4 class ssd		# do not change unnecessarily
	# weight 0.545
	alg straw
	hash 0	# rjenkins1
	item osd.3 weight 0.272
	item osd.0 weight 0.272
}
host ip-10-0-163-134 {
	id -5		# do not change unnecessarily
	id -6 class ssd		# do not change unnecessarily
	# weight 0.545
	alg straw
	hash 0	# rjenkins1
	item osd.2 weight 0.272
	item osd.5 weight 0.272
}
host ip-10-0-144-122 {
	id -7		# do not change unnecessarily
	id -8 class ssd		# do not change unnecessarily
	# weight 0.545
	alg straw
	hash 0	# rjenkins1
	item osd.1 weight 0.272
	item osd.4 weight 0.272
}
root default {
	id -1		# do not change unnecessarily
	id -2 class ssd		# do not change unnecessarily
	# weight 1.635
	alg straw
	hash 0	# rjenkins1
	item ip-10-0-139-184 weight 0.545
	item ip-10-0-163-134 weight 0.545
	item ip-10-0-144-122 weight 0.545
}

# rules
rule replicated_ruleset {
	id 0
	type replicated
	min_size 1
	max_size 10
	step take default
	step chooseleaf firstn 0 type host
	step emit
}

# end crush map
```

Example ceph.conf from an OSD pod, note the absence of zone or region in the crush_location parameter:
```
sh-4.2# cat /etc/ceph/ceph.conf
[global]
fsid                      = 46258bc3-62c5-4c59-b06d-51e5bda56cb0
run dir                   = /var/lib/rook/osd0
mon initial members       = a b c
mon host                  = v1:172.30.164.205:6789,v1:172.30.201.105:6789,v1:172.30.10.98:6789
public addr               = 10.130.4.12
cluster addr              = 10.130.4.12
mon keyvaluedb            = rocksdb
mon_allow_pool_delete     = true
mon_max_pg_per_osd        = 1000
debug default             = 0
debug rados               = 0
debug mon                 = 0
debug osd                 = 0
debug bluestore           = 0
debug filestore           = 0
debug journal             = 0
debug leveldb             = 0
filestore_omap_backend    = rocksdb
osd pg bits               = 11
osd pgp bits              = 11
osd pool default size     = 1
osd pool default min size = 1
osd pool default pg num   = 100
osd pool default pgp num  = 100
osd objectstore           = filestore
crush location            = root=default host=ip-10-0-139-184
rbd_default_features      = 3
fatal signal handlers     = false

[osd.0]
keyring              = /var/lib/ceph/osd/ceph-0/keyring
bluestore block path = /var/lib/ceph/osd/ceph-0/block
```

### PR DESCRIPTION
**Pull request by rohantmp**
Signed-off-by: Rohan CJ <rohantmp@gmail.com>

**Description of your changes:**

The crush location was being stored in `c.ValidStorage.Nodes`, but the value was being stored in `c.DesiredStorage.Nodes`. I ensured that DesiredStorage was being updated as well.

I have verified that the failure-domain labels are being read and propagated to the crush map.

**Which issue is resolved by this Pull Request:**
Resolves #3430

**Checklist:**

- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/v1.0/development-flow.html#submitting-a-pull-request)
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.
- [x] Pending release notes updated with breaking and/or notable changes, if necessary.
- [x] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [x] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [x] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../CONTRIBUTING.md#comments)
- [x] Add the flag for skipping the CI if this PR does not require a build. See [here](/INSTALL.md#skip-ci) for more details.

### CODE REVIEW

File: pkg/apis/rook.io/v1alpha2/storage.go [hunk 1]
```diff
@@ -97,7 +97,6 @@ func (s *StorageScopeSpec) resolveNodeSelection(node *Node) {
 }
 
 func (s *StorageScopeSpec) resolveNodeConfig(node *Node) {
-	resolveString(&(node.Location), s.Location, "")
```
Comment (jarrpa) [inline pkg/apis/rook.io/v1alpha2/storage.go]: Why is this removed?

File: pkg/operator/k8sutil/node.go [hunk 2]
```diff
@@ -84,8 +88,10 @@ func GetValidNodes(rookStorage rookalpha.StorageScopeSpec, clientset kubernetes.
 			validK8sNodes = append(validK8sNodes, n)
 		}
 	}
+	// Get the corresponding rookv1alpha2.Nodes and update them.
+	rookNodes, validNodes := RookNodesMatchingKubernetesNodes(rookStorage, validK8sNodes)
 
-	return RookNodesMatchingKubernetesNodes(rookStorage, validK8sNodes)
+	return rookNodes, validNodes
```
Comment (jarrpa) [inline pkg/operator/k8sutil/node.go]: Why are you returning `rookStorage.Nodes`? I don't see it used anywhere where this function is called.

File: pkg/operator/k8sutil/node.go [hunk 3]
```diff
@@ -237,25 +243,27 @@ func GetKubernetesNodesMatchingRookNodes(rookNodes []rookalpha.Node, clientset k
...
+					rookStorage.Nodes[rnNum].Location = nodeTopologyLocation(kn, rookStorage.Nodes[rnNum].Location)
 				}
 
-				nodes = append(nodes, rn)
+				matchingNodes = append(matchingNodes, rookStorage.Nodes[rnNum])
 			}
 		}
 	}
-	return nodes
+	return rookStorage.Nodes, matchingNodes
```
Comment (jarrpa) [inline pkg/operator/k8sutil/node.go]: Why are you returning `rookStorage.Nodes`? I don't see it used anywhere where this function is called.

File: pkg/operator/ceph/cluster/osd/osd.go [hunk 4]
```diff
@@ -145,6 +145,9 @@ func (c *Cluster) Start() error {
 	if c.DesiredStorage.UseAllNodes == false && len(c.DesiredStorage.Nodes) == 0 {
 		logger.Warningf("useAllNodes is set to false and no nodes are specified, no OSD pods are going to be created")
 	}
+	if len(c.DesiredStorage.Location) > 0 {
+		logger.Warning("Crush Location has been specified on the cluster level storage selection(as opposed to node level). Ignoring! ")
```
[the remark this answers is not in the document]
Comment (rohantmp) [reply pkg/operator/ceph/cluster/osd/osd.go]: +1

File: pkg/apis/rook.io/v1alpha2/storage.go (hunk 1 above)
Comment (rohantmp) [reply pkg/apis/rook.io/v1alpha2/storage.go]: This is the part that's being ignored. This is pulling in the cluster level Location of the StorageScopeSpec.

File: pkg/operator/k8sutil/node.go (hunk 2 above)
Comment (rohantmp) [reply pkg/operator/k8sutil/node.go]: Actually, they're both being used later on, and this is the only entry point for matching them with their k8s counterparts. (worth a refactor later, but the library is used in a lot of places). Somewhere along the callchain it's the Location of `rookStorage.Nodes` that's being used for crush.

File: pkg/operator/k8sutil/node.go (hunk 3 above)
Comment (rohantmp) [reply pkg/operator/k8sutil/node.go]: It's updating the c.DesiredStorage.Nodes of the osd cluster struct.

File: pkg/operator/k8sutil/node.go (hunk 2 above)
Comment (rohantmp) [reply pkg/operator/k8sutil/node.go]: I did try just changing the everything to use ValidStorage, but that seemed to break something and the OSDs deployments wouldn't get created. I think we definitely need a study/refactor of all the assumptions of where the data will be and what triggers what.

File: pkg/apis/rook.io/v1alpha2/storage.go (hunk 1 above)
Comment (travisn) [reply pkg/apis/rook.io/v1alpha2/storage.go]: So we will only look at the `location` added at a lower scope such as individual nodes? I don't know why anyone would even specify the cluster-level location though, so it makes sense.

File: pkg/operator/k8sutil/node.go [hunk 5]
```diff
@@ -65,14 +65,18 @@ func ValidNode(node v1.Node, placement rookalpha.Placement) (bool, error) {
 	return ValidNodeNoSched(node, placement)
 }
 
-// GetValidNodes returns all nodes that (1) are not cordoned, (2) meet Rook's placement terms, and
-// (3) are ready.
-func GetValidNodes(rookStorage rookalpha.StorageScopeSpec, clientset kubernetes.Interface, placement rookalpha.Placement) []rookalpha.Node {
+// GetValidNodes returns rookNodes, validNodes
+// rookNodes are rookStorage.Nodes updated to better reflect the corresponding k8s objects:
```
Comment (travisn) [inline pkg/operator/k8sutil/node.go]: How about we call these `desiredNodes` and `validDesiredNodes`? I'm getting confused with the more general terms.

File: pkg/operator/k8sutil/node.go (hunk 3 above)
Comment (jarrpa) [reply pkg/operator/k8sutil/node.go]: Sure, but why? It wasn't being updated before, what do we gain from updating it now?

File: pkg/operator/k8sutil/node.go (hunk 5 above)
Comment (jarrpa) [reply pkg/operator/k8sutil/node.go]: Agreed, take this chance to rename these to something more obvious.

File: pkg/operator/k8sutil/node.go (hunk 3 above)
Comment (rohantmp) [reply pkg/operator/k8sutil/node.go]: The location is take from [there](https://github.com/rook/rook/blob/c1e5a46887ffb6b31597b7920a748c2c34302067/pkg/operator/ceph/cluster/osd/osd.go#L551)

File: pkg/operator/k8sutil/node.go (hunk 5 above)
Comment (rohantmp) [reply pkg/operator/k8sutil/node.go]: +1

Comment (rohantmp) [reply pkg/operator/k8sutil/node.go]: Done

**Comment by rohantmp** [comment]
I see that the test `go.test.unit` failed. It's failing when I run the tests locally too, but it doesn't say which test failed.

File: pkg/apis/rook.io/v1alpha2/storage_test.go [hunk 6]
```diff
@@ -48,7 +48,6 @@ func TestResolveNodeDefaultValues(t *testing.T) {
 func TestResolveNodeInherentFromCluster(t *testing.T) {
 	// a node with no properties defined should inherit them from the cluster storage spec
 	storageSpec := StorageScopeSpec{
-		Location: "root=default,row=a,rack=a2,chassis=a2a,host=a2a1",
```
Comment (rohantmp) [inline pkg/apis/rook.io/v1alpha2/storage_test.go]: Made these changes to the test because we're ignoring cluster level information info.

Comment (rohantmp) [reply pkg/apis/rook.io/v1alpha2/storage_test.go]: Could this impact other rook inhabitants?

Comment (rohantmp) [reply pkg/apis/rook.io/v1alpha2/storage_test.go]: I see that only EdgeFS uses ResolveNode currently, and they do not use the Location field of v1alpha2.Node.

File: pkg/operator/k8sutil/node.go (hunk 3 above)
Comment (jarrpa) [reply pkg/operator/k8sutil/node.go]: That doesn't tell me why we're updating it NOW (in this PR) when we weren't updating it before. Does this fix something that was previously broken? From the code you linked, the only thing that's used are the resource requirements on the nodes, which are not being modified or referenced here.

Comment (rohantmp) [reply pkg/operator/k8sutil/node.go]: It was the main thing that prevented the crush location from propagating.
GetValidNodes used to return the validNodes (now valid desired nodes) with two changes: 

- node.Name value changed to the node's hostname label.
- node.Location populated with failure domain labels.

So it was doing more than just returning a subset of the nodes, it was updating them with info from the k8s objects.

This info was put in c.ValidStorage.Nodes, but it was also needed in c.DesiredStorage.Nodes. It is also not possible to copy them later because the node.Name in c.ValidStorage was not necessarily the name.

Comment (jarrpa) [reply pkg/operator/k8sutil/node.go]: It seems very wrong that we should be getting any node information from `c.DesiredStorage`, Rook-Ceph should only be using `c.ValidStorage`, just from the name alone. It seems like a bug that `resolveNode()` uses `c.DesiredStorage`. Is it possible to fix that and see if you can get away with only returning/updating `c.ValidStorage`?

Comment (jarrpa) [reply pkg/operator/k8sutil/node.go]: @travisn I'd also like to get your input here, to see if I'm just getting this all wrong.

Comment (travisn) [reply pkg/operator/k8sutil/node.go]: The DesiredStorage is supposed to represent the static settings in the CR, while the ValidStorage represents what was discovered. In the past the `location` was only assumed to be on the CR, but now we're discovering it dynamically. It really should be combined into the ValidStorage where dynamic info is used, so it does look like we need to change the location handling to use ValidStorage instead of DesiredStorage.

File: pkg/operator/k8sutil/node.go [hunk 7]
```diff
@@ -65,14 +65,18 @@ func ValidNode(node v1.Node, placement rookalpha.Placement) (bool, error) {
...
-// GetValidNodes returns all nodes that (1) are not cordoned, (2) meet Rook's placement terms, and
-// (3) are ready.
-func GetValidNodes(rookStorage rookalpha.StorageScopeSpec, clientset kubernetes.Interface, placement rookalpha.Placement) []rookalpha.Node {
+// GetValidNodes returns desiredNodes, validDesiredNodes
+// desiredNodes are rookStorage.Nodes updated to better reflect the corresponding k8s objects:
+// validDesiredNodes are a subset of desiredNodes that are:
+//     - are not cordoned
+//     - meet Rook's placement terms
+//     - are Ready.
+func GetValidNodes(rookStorage rookalpha.StorageScopeSpec, clientset kubernetes.Interface, placement rookalpha.Placement) ([]rookalpha.Node, []rookalpha.Node) {
```
Comment (BlaineEXE) [inline pkg/operator/k8sutil/node.go]: I would like to see the changes to this file reverted entirely. This pretty clearly seems to be a change intended only for Ceph, but by changing this file, it is changing behavior for any backend. And, this is breaking the rules of single-purpose functions. 

It seems that the StorageScopeSpec may be somewhat broken if it needs to be updated to better reflect the k8s objects. Maybe a better way would be to write a separate function similar to the ones used to fully resolve nodes.

File: pkg/operator/k8sutil/node.go (hunk 3 above)
Comment (BlaineEXE) [reply pkg/operator/k8sutil/node.go]: I agree with @jarrpa that this doesn't fully make sense. It feels hacky at best.

Comment (BlaineEXE) [reply pkg/operator/k8sutil/node.go]: If you look at the comment where `DesiredStorage` is defined, it is this
```go
DesiredStorage  rookalpha.StorageScopeSpec // user-defined storage scope spec
```

DesiredStorage is intended to be a read-only structure containing the user's original requests. This should **never** be modified.

**Comment by rohantmp** [comment]
Figured out that the reason switching to ValidStorage wasn't working was because only ValidStorage.Nodes was being initialized. 

All the other changes were unecessary... No need to drop cluster level Location because it is only populated if node level Location is not set even though it doesn't make sense to use it in Ceph's case.

**Comment by rohantmp** [comment]
Done!
